hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/TimelineService.java

```diff
@@ -342,8 +341,19 @@ private int startServiceOnPort(int port) throws IOException {
   }
 
   public int startService() throws IOException {
-    final Server server = timelineServerConf.numThreads == DEFAULT_NUM_THREADS ? new JettyServer(new JavalinConfig()).server() :
-            new Server(new QueuedThreadPool(timelineServerConf.numThreads));
+    int maxThreads;
+    if (timelineServerConf.numThreads > 0) {
+      maxThreads = timelineServerConf.numThreads;
+    } else {
+      // io.javalin.jetty.JettyUtil.defaultThreadPool
+      maxThreads = 250;
+    }
+    QueuedThreadPool pool = new QueuedThreadPool(maxThreads, 8, 60_000);
+    pool.setDaemon(true);
+    final Server server = new Server(pool);
+    ScheduledExecutorScheduler scheduler = new ScheduledExecutorScheduler("TimelineService-JettyScheduler", true, 8);
+    server.addBean(scheduler);
+
```

Comment:
Not sure what the effect is for `server.addBean(scheduler)`, and when the scheduler is shut down?

Reply:
Both `QueuedThreadPool` and `ScheduledExecutorScheduler` inherit `AbstractLifeCycle`.
When the Server stops, it will stop all the beans that implement `LifeCycle`.

org.eclipse.jetty.server.Server#doStop

org.eclipse.jetty.util.component.ContainerLifeCycle#doStop

Reply:
So the `#addBean` is used for lifecycle management of the server components.